td-layout/src/lib.rs

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) 2020 Intel Corporation
+//
+// SPDX-License-Identifier: BSD-2-Clause-Patent
+
+#![no_std]
+
+//! Define file (build-time) and runtime layout for shim binary.
+//!
+//! Note:
+//! `repr(C)` should be used to control the exact data structure layout.
+//! - `repr(rust)`: By default, composite structures have an alignment equal to the maximum of their
+//!   fields' alignments. Rust will consequently insert padding where necessary to ensure that all
+//!   fields are properly aligned and that the overall type's size is a multiple of its alignment.
+//!   And fields may be reordered, not following the literal order of fields.
+//! - `repr(C)` is the most important repr. It has fairly simple intent: do what C does. The order,
+//!   size, and alignment of fields is exactly what you would expect from C or C++. Any type you
+//!   expect to pass through an FFI boundary should have repr(C), as C is the lingua-franca of the
+//!   programming world. This is also necessary to soundly do more elaborate tricks with data layout
+//!   such as reinterpreting values as a different type.
+
+use core::fmt;
+
+pub mod build_time;
+pub mod mailbox;
+pub mod memslice;
+pub mod metadata;
+pub mod runtime;
+
+// Minimal memory size to build the runtime layout.
+const MIN_MEMORY_SIZE: u64 = 0x3000000;
+
+#[derive(Default)]
+pub struct RuntimeMemoryLayout {
+    pub runtime_page_table_base: u64,
+    pub runtime_payload_param_base: u64,
+    pub runtime_payload_base: u64,
+
+    pub runtime_event_log_base: u64,
+    pub runtime_acpi_base: u64,
+    pub runtime_hob_base: u64,
+    pub runtime_shadow_stack_base: u64,
+    pub runtime_stack_base: u64,
+    pub runtime_heap_base: u64,
+    pub runtime_dma_base: u64,
+
+    pub runtime_stack_top: u64,
+    pub runtime_shadow_stack_top: u64,
+
+    pub runtime_memory_bottom: u64,
+}
+
+impl RuntimeMemoryLayout {
+    pub fn new(memory_top: u64) -> Self {
+        use crate::runtime::*;
+        let current_base = memory_top & !0xfffff;
+
+        if memory_top < MIN_MEMORY_SIZE {
+            panic!("memory_top 0x{:x} is too small", memory_top);
+        }
+
+        let current_base = current_base - TD_PAYLOAD_EVENT_LOG_SIZE as u64;
```

Comment:
Given that you almost always convert the values to `u64`, why define the constants as `u32` to start with?

Reply:
Did a quick experiment, seems to work fine, see https://github.com/c3d/td-shim/commit/6d8aa66e809c98d25ad094f10bf6140dad7970de.

Reply:
Great, how about opening a new on top of this one?

Reply:
One reason to use u32 instead of u64 it that td-shim always laid below 4G, and these constants may be converted to u64/usize. For 64 bit platforms, usize is the same as u64 though:)

Reply:
> One reason to use u32 instead of u64 it that td-shim always laid below 4G, and these constants may be converted to u64/usize. For 64 bit platforms, usize is the same as u64 though:)

I understand that. My comment is more with respect to the Rust type system. I find it ugly to have `constant as u64` repeated all over the place, this could hide actual issues.